File: 0020-valid-parentheses/0020-valid-parentheses.js
/**
 * @param {string} s
 * @return {boolean}
 */
var isValid = function(s) {
    var stack = [];
    for (let i = 0; i < s.length; i++) {
        var nextChar = s.charAt(i);
        if (nextChar === ')' || nextChar === ']' || nextChar === '}') {
            var topOfStack = stack.pop();
            if (topOfStack !== getCharComplement(nextChar)) {
                return false;
            }
        } else {
            stack.push(nextChar);
        }
    }
    return !stack.length;
};
    
var getCharComplement = (ch) => {
    switch (ch) {
        case ')':
            return '(';
        case ']':
            return '[';
        case '}':
            return '{';
        default:
            return undefined;    
    }
}

File: 0020-valid-parentheses/0020-valid-parentheses.py
class Solution:
    def isValid(self, s: str) -> bool:
        d = {
            ")": "(",
            "}": "{",
            "]": "["
        }
        
        openers = []
        
        for c in s:
            if c in d.values():
                openers.append(c)
            elif len(openers) > 0 and d[c] == openers[-1]:
                openers.pop()
            else:
                return False
            
        return len(openers) == 0